File: 098*validate-binary-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.prev = None
        
    def isValidBST(self, root: TreeNode) -> bool:
        if not root:
            return True
        
        if self.isValidBST(root.left) is False:
            return False
        
        if self.prev is None:
            self.prev = root.val
        else:
            if root.val <= self.prev:
                return False
            else:
                self.prev = root.val
            
        if self.isValidBST(root.right) is False:
            return False
        
        return True

File: 098*validate-binary-search-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
import (
    "math"
    "fmt"
)


func auxIsValidBST(root *TreeNode, prev *int) bool {
    if root == nil {
        return true
    }
    
    if !auxIsValidBST(root.Left, prev) {
        return false
    } 
     
    if root.Val <= *prev {
        return false
    } else{
       *prev = root.Val     
    }

    if !auxIsValidBST(root.Right, prev) {
        return false
    }
    
    return true  
}

func isValidBST(root *TreeNode) bool {
    var prev int = int(math.Inf(-1))
    return auxIsValidBST(root, &prev)
    
}
